### PR TITLE
New features in LDT for GVF CDF generation and SMAP_E_OPL retrievals

### PR DESCRIPTION
**Pull request by yhkwon81**
### Description

Codes for VIIRS and CDFS CDF generation and operational enhanced SMAP (SMAP_E_OPL) soil moisture retrievals have been added in LDT.

### Testcase
SMAP_E_OPL: /discover/nobackup/projects/usaf_lis/ykwon/SMAP_E_OPL/testcases/TEST_01_usecase
VIIRS CDF: /discover/nobackup/projects/usaf_lis/ykwon/SoilMoisture_DA/testcases/noahmp401/noahmp401_CONUS_10km_dynamic/DA_proc_VIIRS
CDFS CDF: /discover/nobackup/projects/usaf_lis/ykwon/SoilMoisture_DA/testcases/noahmp401/noahmp401_CONUS_10km_dynamic/DA_proc_CDFS

### CODE REVIEW

**Comment by yhkwon81** [comment]
Hi Brendan,

Thanks a lot for checking these.

Regarding the LDT codes, I fixed “LDT_DAobsDataMod.F90” and could compile LDT with “-2” mode.


  *   VIIRS CDF: the testcase directory contains a soft link to a VIIRS_CDF/ directory, but the target doesn't exist (/discover/nobackup/projects/usaf_lis/dmocko/gvf/VIIRS). Can you point me to another copy of these inputs?  • I updated the directory link. Could you check it again?
  *   SMAP_E_OPL: the ldt.config file points to /discover/nobackup/projects/usaf_lis/ykwon/SMAP_E_OPL/data/SMAP_L1B_historical, but I do not have read access on this directory.
• I changed the permission. Could you try to access it now?

Thanks,
Yonghwan


From: Brendan McAndrew ***@***.***>
Reply-To: NASA-LIS/LISF ***@***.***>
Date: Tuesday, June 7, 2022 at 9:26 PM
To: NASA-LIS/LISF ***@***.***>
Cc: "Kwon, Yonghwan (GSFC-617.0)[UNIVERSITY OF MARYLAND]" ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [NASA-LIS/LISF] New features in LDT for GVF CDF generation and SMAP_E_OPL retrievals (PR #1110)


Hi again,

I am also unable to access input files needed for two of the testcases:

  *   VIIRS CDF: the testcase directory contains a soft link to a VIIRS_CDF/ directory, but the target doesn't exist (/discover/nobackup/projects/usaf_lis/dmocko/gvf/VIIRS). Can you point me to another copy of these inputs?
  *   SMAP_E_OPL: the ldt.config file points to /discover/nobackup/projects/usaf_lis/ykwon/SMAP_E_OPL/data/SMAP_L1B_historical, but I do not have read access on this directory.

**Comment by bmcandr** [comment]
Thanks @yhkwon81, I am able to access those files now.

**Comment by bmcandr** [comment]
I'm now getting an error on the SMAP_E_OPL testcase:

```
forrtl: severe (408): fort: (2): Subscript #1 of the array ZERODISTFLAG has value 2555 which is greater than the upper bound of 1920

Image              PC                Routine            Line        Source
LDT.intel.21.defa  0000000000BF069F  Unknown               Unknown  Unknown
LDT.intel.21.defa  00000000008A1771  invdist_l1b2arfs_         121  invdist_l1b2arfs.F90
LDT.intel.21.defa  00000000006E6CAE  smapl1bresample_           63  SMAPL1BTOL1C_ARFS.F90
LDT.intel.21.defa  0000000000640125  ldt_smap_e_oplmod         272  LDT_smap_e_oplMod.F90
LDT.intel.21.defa  000000000063CC89  ldt_run_smap_e_op          25  LDT_run_smap_e_opl.F90
LDT.intel.21.defa  000000000063D198  Unknown               Unknown  Unknown
LDT.intel.21.defa  0000000000681059  MAIN__                     46  LDTmain.F90
LDT.intel.21.defa  0000000000414352  Unknown               Unknown  Unknown
libc-2.22.so       00002AAAB46F4725  __libc_start_main     Unknown  Unknown
LDT.intel.21.defa  0000000000414269  Unknown               Unknown  Unknown
```

I'm wondering if the array `zerodistflag` is being allocated with the dimensions swapped?

https://github.com/NASA-LIS/LISF/blob/59ea4a0f9d233f4b1a6bc22f1a0ec995c387ac2f/ldt/SMAP_E_OPL/invdist_l1b2arfs.F90#L39

File: ldt/SMAP_E_OPL/SMAPL1BTOL1C_ARFS.F90 [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+!
+! DESCRIPTION: RESAMPLE SMAPL1B TB TO AIR FORCE GRID
+! INPUT : SMAP - L1B Brightness Temperature               
+! OUTPUT: SMAPTB_ARFSGRIDE_ddmmyyy.dat
+! NOTES : Inverse Distance Squared with 0.4 deg serching window
+!-------------------------------------------------------------------------
+
+subroutine SMAPL1BRESAMPLE(SMAPFILE,L1B_dir,Orbit,ARFS_TIME)
+
+   USE HDF5
```
Comment (bmcandr) [inline ldt/SMAP_E_OPL/SMAPL1BTOL1C_ARFS.F90:19]: Remove `USE HDF5`. Doesn't look like it is used in this subroutine.

File: ldt/SMAP_E_OPL/TOOLSUBS.F90 [hunk 2]
```diff
@@ -0,0 +1,382 @@
+!=============================i==========================================
+!  MODULE, TOOLSUBS P.W.LIU, 09/12/18
+!  Contains subroutines that are necessry for downscaling program
+!-----------------------------------------------------------------------
+!  AVHRR_NDVI
+!  NOTES : Modified P.W. Liu, 09/12/18
+!=======================================================================
+MODULE TOOLSUBS
+    USE FUNCTIONS
+    USE HDF5
```
Comment (bmcandr) [inline ldt/SMAP_E_OPL/TOOLSUBS.F90:10]: Wrap in `#if (defined USE_HDF5)` preprocessor check.

File: ldt/SMAP_E_OPL/TOOLSUBS.F90 [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+!       PRINT*,'ds_id hdferr', dataset_id, hdferr
+
+        ALLOCATE(data_out(smap_row,smap_col),sm_mat(size(lon_ind),size(lat_ind)))
+        data_dims(1) = smap_row
+        data_dims(2) = smap_col
+        CALL h5dread_f(dataset_id, H5T_NATIVE_DOUBLE, data_out, data_dims, hdferr) !Read dataset
+        CALL h5dclose_f(dataset_id,hdferr)
+        CALL h5fclose_f(file_id,hdferr)
+        CALL h5close_f(hdferr)
+        sm_mat = data_out(lon_ind(1):lon_ind(size(lon_ind)),lat_ind(1):lat_ind(size(lat_ind)))
```
Comment (bmcandr) [inline ldt/SMAP_E_OPL/TOOLSUBS.F90:122]: Wrap in `#if (defined USE_HDF5)` preprocessor check.

File: ldt/SMAP_E_OPL/TOOLSUBS.F90 [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+        CALL h5dread_f(dataset_id3, H5T_NATIVE_INTEGER, row_temp_ind, dims, hdferr) !Read dataset
+        ind=row_temp_ind*3856+col_temp_ind+1 !2D Matrix index to 1D vector index
+!        print *, size(data_out), data_out(192571), data_out(205664)
+!        print *, row_temp_ind(192571), row_temp_ind(205664)
+!        print *, col_temp_ind(192571), col_temp_ind(205664)
+        CALL h5dclose_f(dataset_id1,hdferr)
+!        CALL h5dclose_f(dataset_id2,hdferr)
+!        CALL h5dclose_f(dataset_id3,hdferr)
+        CALL h5fclose_f(file_id,hdferr)
+        CALL h5close_f(hdferr)
```
Comment (bmcandr) [inline ldt/SMAP_E_OPL/TOOLSUBS.F90:159]: Wrap in `#if (defined USE_HDF5)` preprocessor check.

File: ldt/SMAP_E_OPL/TOOLSUBS.F90 [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+        CALL h5dclose_f(dataset_id8,hdferr)
+        CALL h5dclose_f(dataset_id9,hdferr)
+        CALL h5dclose_f(dataset_id10,hdferr)
+        CALL h5dclose_f(dataset_id11,hdferr)
+        CALL h5dclose_f(dataset_id12,hdferr)
+        CALL h5dclose_f(dataset_id13,hdferr)
+        CALL h5dclose_f(dataset_id14,hdferr)
+        CALL h5dclose_f(dataset_id15,hdferr)
+        CALL h5fclose_f(file_id,hdferr)
+        CALL h5close_f(hdferr)
```
Comment (bmcandr) [inline ldt/SMAP_E_OPL/TOOLSUBS.F90:264]: Wrap in `#if (defined USE_HDF5)` preprocessor check.

**Review by bmcandr — CHANGES_REQUESTED** [review]
Hi @yhkwon81,

I am requesting a few more changes before we can merge this in. See the comments on the pull request in GitHub.

**Comment by emkemp** [comment]
Hi @yhkwon81 

I dusted off this pull request, and there are issues that need to be addressed.

* Documentation needs to be added for new ldt.config entries.
* Parts of the code need to be wrapped in #if (defined USE_HDF5) and #endif, as HDF5 is not linked by default.

In addition, I cannot replicate the results for the VIIRS CDF nor the CDFS CDF.  In both cases LDT fails to read the GVF files, despite my ldt.config files looking very much like your samples.  I have not determined the cause yet.

Until these changes are made, the pull request will remained marked as Not Ready.

**Comment by yhkwon81** [comment]
Thanks for reminding me these. I will resolve the issues.

From: Eric Kemp ***@***.***>
Reply-To: NASA-LIS/LISF ***@***.***>
Date: Friday, July 22, 2022 at 1:55 PM
To: NASA-LIS/LISF ***@***.***>
Cc: "Kwon, Yonghwan (GSFC-617.0)[UNIVERSITY OF MARYLAND]" ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [NASA-LIS/LISF] New features in LDT for GVF CDF generation and SMAP_E_OPL retrievals (PR #1110)


Hi @yhkwon81<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fyhkwon81&data=05%7C01%7Cyonghwan.kwon%40nasa.gov%7Ceaf8b510679c46e62b0a08da6c0b55e5%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637941093204992865%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=lx%2F8i%2F%2FzmTOsxxjDbrLRa44jCZXMOYDraj%2BtDinYw1E%3D&reserved=0>

I dusted off this pull request, and there are issues that need to be addressed.

  *   Documentation needs to be added for new ldt.config entries.
  *   Parts of the code need to be wrapped in #if (defined USE_HDF5) and #endif, as HDF5 is not linked by default.

In addition, I cannot replicate the results for the VIIRS CDF nor the CDFS CDF. In both cases LDT fails to read the GVF files, despite my ldt.config files looking very much like your samples. I have not determined the cause yet.

Until these changes are made, the pull request will remained marked as Not Ready.

**Comment by emkemp** [comment]
Thanks @yhkwon81 

Update on my end:  I figured out why the VIIRS and CDFS files weren't being read in (silly user error).  I've started rerunning for comparison with your test case.

Also, I notice that for the SMAP case, 'diff' indicates my ARFS_SM_V_01798_D_20150603T122325_R17000_001.dat is not the same as yours.  Is it possible to plot these files and look at the differences?

**Comment by yhkwon81** [comment]
Thanks for the update. Please let me know if you encounter any further issues.
Could you direct me to your SMAP case directory?

From: Eric Kemp ***@***.***>
Reply-To: NASA-LIS/LISF ***@***.***>
Date: Friday, July 22, 2022 at 2:40 PM
To: NASA-LIS/LISF ***@***.***>
Cc: "Kwon, Yonghwan (GSFC-617.0)[UNIVERSITY OF MARYLAND]" ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [NASA-LIS/LISF] New features in LDT for GVF CDF generation and SMAP_E_OPL retrievals (PR #1110)


Thanks @yhkwon81<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fyhkwon81&data=05%7C01%7Cyonghwan.kwon%40nasa.gov%7C2314416cb6f44cb75f4208da6c119a53%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637941120107915713%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=9r%2FAspYnxE%2BDlUpsPn0it7lPlpqPyF0%2FwwA6syu36L8%3D&reserved=0>

Update on my end: I figured out why the VIIRS and CDFS files weren't being read in (silly user error). I've started rerunning for comparison with your test case.

Also, I notice that for the SMAP case, 'diff' indicates my ARFS_SM_V_01798_D_20150603T122325_R17000_001.dat is not the same as yours. Is it possible to plot these files and look at the differences?

**Comment by emkemp** [comment]
Sure, it's /discover/nobackup/emkemp/pr/1110/work_sm

Read permissions *should* be set for your access.

**Comment by yhkwon81** [comment]
Thanks. I think that my results were generated before fixing some bugs in LDT. I re-generated the SMAP_E_OPL soil moisture using the bug-fixed LDT, and “diff” shows that yours and mine are the same. My new results are in the same directory (/discover/nobackup/projects/usaf_lis/ykwon/SMAP_E_OPL/testcases/TEST_01_usecase/SM_retrieval).


From: Eric Kemp ***@***.***>
Reply-To: NASA-LIS/LISF ***@***.***>
Date: Friday, July 22, 2022 at 3:06 PM
To: NASA-LIS/LISF ***@***.***>
Cc: "Kwon, Yonghwan (GSFC-617.0)[UNIVERSITY OF MARYLAND]" ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [NASA-LIS/LISF] New features in LDT for GVF CDF generation and SMAP_E_OPL retrievals (PR #1110)


Sure, it's /discover/nobackup/emkemp/pr/1110/work_sm

Read permissions should be set for your access.

**Comment by emkemp** [comment]
Hi @yhkwon81 

Can you open up read permissions to /discover/nobackup/projects/usaf_lis/ykwon/SMAP_E_OPL/testcases/TEST_01_usecase/SM_retrieval?  I'm locked out.

**Comment by yhkwon81** [comment]
I am sorry. I changed the permission. Please try again.

From: Eric Kemp ***@***.***>
Reply-To: NASA-LIS/LISF ***@***.***>
Date: Thursday, July 28, 2022 at 5:34 PM
To: NASA-LIS/LISF ***@***.***>
Cc: "Kwon, Yonghwan (GSFC-617.0)[UNIVERSITY OF MARYLAND]" ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [NASA-LIS/LISF] New features in LDT for GVF CDF generation and SMAP_E_OPL retrievals (PR #1110)


Hi @yhkwon81<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fyhkwon81&data=05%7C01%7Cyonghwan.kwon%40nasa.gov%7Cfbd055c05d40473088ad08da70e0eb7d%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637946408573457548%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=4iYkPCACuPEMqY7G4wAAje4zv9ff%2BA8NY6DGyU2ca%2FM%3D&reserved=0>

Can you open up read permissions to /discover/nobackup/projects/usaf_lis/ykwon/SMAP_E_OPL/testcases/TEST_01_usecase/SM_retrieval? I'm locked out.

**Comment by emkemp** [comment]
Thanks @yhkwon81!  I confirm I get identical results for ARFS_SM_V_01798_D_20150603T122325_R17000_001.dat.

I added some HDF5 related preprocessing statements to ldt/SMAP_E_OPL/TOOLSUBS.F90.  I will commit to the branch.

**Comment by emkemp** [comment]
Hi @yhkwon81 

I reproduced the netCDF output files for both the VIIRS and CDFS test cases.

You still need to add documentation for the new ldt.config options.  Once that is done, I will coordinate with @jvgeiger on where/how to store your test cases before accepting the pull request.

**Comment by emkemp** [comment]
Hi @yhkwon81 

Thanks for updating the documentation.  At this point I believe the pull request itself is ready.  I want to confer with @jvgeiger on handling your test cases before we proceed.  Fingers crossed, this pull request will be accepted by the end of the week.

**Pull request merged**